crates/wasi-http/src/http_impl.rs

```diff
@@ -0,0 +1,203 @@
+            let connector = tokio_rustls::TlsConnector::from(Arc::new(config));
+            let mut parts = authority.split(":");
+            let host = parts.next().unwrap_or(&authority);
+            let domain =
+                rustls::ServerName::try_from(host).map_err(|_| anyhow!("invalid dnsname"))?;
+            let stream = connector.connect(domain, stream).await?;
+
+            let t = timeout(
+                connect_timeout,
+                hyper::client::conn::http1::handshake(stream),
+            )
+            .await?;
+            let (s, conn) = t?;
+            tokio::task::spawn(async move {
+                if let Err(err) = conn.await {
+                    println!("Connection failed: {:?}", err);
+                }
+            });
+            s
+        } else {
+            let tcp = TcpStream::connect(authority).await?;
+            let t = timeout(connect_timeout, hyper::client::conn::http1::handshake(tcp)).await?;
+            let (s, conn) = t?;
+            tokio::task::spawn(async move {
+                if let Err(err) = conn.await {
+                    println!("Connection failed: {:?}", err);
+                }
+            });
+            s
+        };
+
+        let url = scheme.to_owned() + &request.authority + &request.path + &request.query;
+
+        let mut call = Request::builder()
+            .method(method)
+            .uri(url)
+            .header(hyper::header::HOST, request.authority.as_str());
+
+        for (key, val) in request.headers.iter() {
+            for item in val {
+                call = call.header(key, item.clone());
+            }
+        }
+
+        let response_id = self.response_id_base;
+        self.response_id_base = self.response_id_base + 1;
+        let mut response = ActiveResponse::new(response_id);
+        let body = Full::<Bytes>::new(
+            self.streams
+                .get(&request.body)
+                .unwrap_or(&Bytes::new())
+                .clone(),
+        );
+        let t = timeout(first_bytes_timeout, sender.send_request(call.body(body)?)).await?;
+        let mut res = t?;
+        response.status = res.status().try_into()?;
+        for (key, value) in res.headers().iter() {
+            let mut vec = std::vec::Vec::new();
+            vec.push(value.to_str()?.to_string());
+            response
+                .response_headers
+                .insert(key.as_str().to_string(), vec);
+        }
+        let mut buf = BytesMut::new();
+        while let Some(next) = timeout(between_bytes_timeout, res.frame()).await? {
+            let frame = next?;
+            if let Some(chunk) = frame.data_ref() {
+                buf.put(chunk.clone());
+            }
+        }
+        response.body = self.streams_id_base;
+        self.streams_id_base = self.streams_id_base + 1;
+        self.streams.insert(response.body, buf.freeze());
+        self.responses.insert(response_id, response);
+        Ok(response_id)
```

Comment:
we should actually be returning the future incoming response id here and not the response id

Reply:
Future's aren't implemented (yet) in this PR so the response_id is the same as the "future" id. If you look at the future-incoming-response-get in component_impl.rs, you'll see that it just returns the passed in future as the response id.

https://github.com/bytecodealliance/wasmtime/pull/5929/files/f02db0f19f99eeb82c369b511cb56743f31b7f60#diff-902e7d68ea9ed72ae165cb94771bad9528553b6b1a04661f731bbe40460ff1beR215

Eventually we will implement futures and this will change, but not in this PR.

Reply:
after this is merged, i will submit a pr that will at least prevent the host from crashing if the emulated future is assumed by the guest like here: https://github.com/bytecodealliance/jco/blob/2b3c9cd7839b3728f74af2cbf6722cf09287f956/packages/preview2-shim/lib/http/wasi-http.js#L263-L296